indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java

```diff
@@ -941,9 +939,37 @@ private TaskStatus generateAndPublishSegments(
               ingestionSchema
           );
 
+      Set<DataSegment> tombStones = Collections.emptySet();
+      if (ingestionSchema.getIOConfig().isDropExisting()) {
+        TombstoneHelper tombstoneHelper = new TombstoneHelper(pushed.getSegments(),
+                                                              ingestionSchema.getDataSchema(),
+                                                              toolbox.getTaskActionClient());
+
+        List<Interval> tombstoneIntervals = tombstoneHelper.computeTombstoneIntervals();
+        // now find the versions for the tombstone intervals
+        Map<Interval, String> tombstonesAndVersions = new HashMap<>();
+        for (Interval interval : tombstoneIntervals) {
+          NonnullPair<Interval, String> intervalAndVersion =
+              findIntervalAndVersion(
```

Comment:
Generally speaking, at this point, the time interval should've already been locked, but the method that you are calling here has a bunch of logic that is trying to deal with a lock not yet existing.  Do you have a test case that exercises the code to somehow cause an interval to be hit here that isn't already locked?  I'm curious about the case of events that can lead to that, as there is likely a bug somewhere else in the stack.  This code should be able to just use the locks that already exist to get a version (that version should also generally be the same for all of the locks), it should most likely throw an error if it's gotten to this point and not locked the interval yet.

Reply:
@imply-cheddar At the time that I wrote that code I could not find an easy way to use the allocators (`SegmentAllocator` implementations) since they are typically created and used in the sub-tasks but the current code in this PR is creating them after the sub-tasks completed (in the parallel case). But I since thought of a potential way to integrate them in the sub-tasks (have each sub-task create the tombstones after it completed pushing all segments, create the tombstones there using its allocator, and then propagate the tombstones back to the supervisor task in the task report, then have the supervisor filter out unneeded tombstones). I will work next on using this new path, with the allocators, since the allocator should be already prepared to deal with issues as the one you comment above. So please hold on for a little while while I try to implement this new path and provide an update to this PR.

Reply:
Took back the code where tombstones are created in the sub-tasks because they generate incorrect tombstones. I thought they could be cleaned up/removed at the supervisor task level but attempting that caused missing core partitions some times. Now the allocation is similar as before, straightforward: just use the versions from the locks. For now, code does no support segment locking and if segment locking is requested the code forces time chunk locking. Segment locking support is left as future work.